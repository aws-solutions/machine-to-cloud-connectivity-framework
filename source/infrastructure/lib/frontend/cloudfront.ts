// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { CustomResource, CfnCondition, CfnCustomResource, Aws } from 'aws-cdk-lib';
import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { SecurityPolicyProtocol } from 'aws-cdk-lib/aws-cloudfront';
import { Bucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface CloudFrontConstructProps {
  readonly s3LoggingBucket: IBucket;
  readonly customResourcesFunctionArn: string;
  readonly shouldTeardownData: CfnCondition;
}

/**
 * Creates a CloudFront distribution.
 */
export class CloudFrontConstruct extends Construct {
  public cloudFrontDomainName: string;
  public uiBucket: Bucket;

  constructor(scope: Construct, id: string, props: CloudFrontConstructProps) {
    super(scope, id);

    const cloudFrontToS3 = new CloudFrontToS3(this, 'CloudFrontToS3', {
      bucketProps: {
        serverAccessLogsBucket: props.s3LoggingBucket,
        serverAccessLogsPrefix: 'ui-s3/',
        bucketName: `${Aws.STACK_NAME}-${Aws.ACCOUNT_ID}-ui`
      },
      cloudFrontDistributionProps: {
        comment: 'Machine to Cloud Connectivity Framework Distribution',
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
        ],
        logBucket: props.s3LoggingBucket,
        minimumProtocolVersion: SecurityPolicyProtocol.TLS_V1_2_2019,
        logFilePrefix: 'ui-cf/'
      },
      insertHttpSecurityHeaders: false
    });
    this.cloudFrontDomainName = cloudFrontToS3.cloudFrontWebDistribution.domainName;
    this.uiBucket = <Bucket>cloudFrontToS3.s3Bucket;

    const teardownCloudfrontBucket = new CustomResource(this, 'TeardownCloudfrontBucket', {
      serviceToken: props.customResourcesFunctionArn,
      properties: {
        Resource: 'DeleteS3Bucket',
        BucketName: this.uiBucket.bucketName
      }
    });
    const cfnTeardownCloudfrontBucket = <CfnCustomResource>teardownCloudfrontBucket.node.defaultChild;
    cfnTeardownCloudfrontBucket.cfnOptions.condition = props.shouldTeardownData;

    NagSuppressions.addResourceSuppressions(
      cloudFrontToS3,
      [
        { id: 'AwsSolutions-CFR1', reason: 'The solution does not control geo restriction.' },
        { id: 'AwsSolutions-CFR2', reason: 'No need to enable WAF.' },
        {
          id: 'AwsSolutions-CFR4',
          reason: 'No contorl on the solution side as it is using the CloudFront default certificate.'
        }
      ],
      true
    );
  }
}
