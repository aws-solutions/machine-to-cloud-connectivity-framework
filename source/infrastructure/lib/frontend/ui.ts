// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Duration, RemovalPolicy, Stack } from 'aws-cdk-lib';
import {
  CfnIdentityPool,
  CfnIdentityPoolRoleAttachment,
  CfnUserPool,
  CfnUserPoolUser,
  UserPool,
  UserPoolClient
} from 'aws-cdk-lib/aws-cognito';
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from 'aws-cdk-lib/aws-iam';
import { CfnBucket, IBucket } from 'aws-cdk-lib/aws-s3';
import { NagSuppressions } from 'cdk-nag';
import { Construct } from 'constructs';

export interface UiConstructProps {
  readonly apiId: string;
  readonly resourceBucket: IBucket;
  readonly userEmail: string;
  readonly cloudFrontDomainName: string;
}

/**
 * Creates a CloudFront distribution, an UI hosting S3 bucket, and Cognito resources.
 */
export class UiConstruct extends Construct {
  public identityPoolId: string;
  public userPoolId: string;
  public webClientId: string;

  constructor(scope: Construct, id: string, props: UiConstructProps) {
    super(scope, id);

    const userPool = new UserPool(this, 'UserPool', {
      passwordPolicy: {
        minLength: 12,
        requireDigits: true,
        requireLowercase: true,
        requireSymbols: true,
        requireUppercase: true
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
      userInvitation: {
        emailSubject: '[Machine to Cloud Connectivity Framework] Login information',
        emailBody: `
          <p>
            You are invited to join Machine to Cloud Connectivity Framework.<br />
            https://${props.cloudFrontDomainName}
          </p>
          <p>
            Please sign in to Machine to Cloud Connectivity Framework using the temporary credentials below:<br />
            Username: <strong>{username}</strong><br />Password: <strong>{####}</strong>
          </p>
        `
      }
    });
    (<CfnUserPool>userPool.node.defaultChild).userPoolAddOns = { advancedSecurityMode: 'ENFORCED' };
    this.userPoolId = userPool.userPoolId;

    const userPoolClient = new UserPoolClient(this, 'UserPoolClient', {
      generateSecret: false,
      preventUserExistenceErrors: true,
      refreshTokenValidity: Duration.days(1),
      userPool,
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`
    });
    this.webClientId = userPoolClient.userPoolClientId;

    new CfnUserPoolUser(this, 'User', {
      userPoolId: userPool.userPoolId,
      desiredDeliveryMediums: ['EMAIL'],
      forceAliasCreation: true,
      userAttributes: [
        { name: 'email', value: props.userEmail },
        { name: 'email_verified', value: 'true' }
      ],
      username: props.userEmail
    });

    const identityPool = new CfnIdentityPool(this, 'IdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [
        {
          clientId: userPoolClient.userPoolClientId,
          providerName: userPool.userPoolProviderName,
          serverSideTokenCheck: false
        }
      ]
    });
    this.identityPoolId = identityPool.ref;

    const authenticatedRole = new Role(this, 'IdentityPoolAuthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'authenticated' }
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        ExecuteApiPolicy: new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['execute-api:Invoke'],
              resources: [
                Stack.of(this).formatArn({
                  service: 'execute-api',
                  resource: props.apiId,
                  resourceName: 'prod/*'
                })
              ]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['s3:GetObject'],
              resources: [props.resourceBucket.arnForObjects('public/*')]
            })
          ]
        })
      }
    });

    new CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoleAttachment', {
      identityPoolId: identityPool.ref,
      roles: { authenticated: authenticatedRole.roleArn }
    });

    (props.resourceBucket.node.defaultChild as CfnBucket).corsConfiguration = {
      corsRules: [
        {
          allowedMethods: ['GET'],
          allowedOrigins: [`https://${props.cloudFrontDomainName}`],
          allowedHeaders: ['*'],
          exposedHeaders: ['ETag']
        }
      ]
    };

    NagSuppressions.addResourceSuppressions(userPool, [
      { id: 'AwsSolutions-COG2', reason: 'No need to enable MFA as that is option for users.' }
    ]);
    NagSuppressions.addResourceSuppressions(authenticatedRole, [
      { id: 'AwsSolutions-IAM5', reason: 'It does not allow wildcard permissions.' }
    ]);
  }
}
