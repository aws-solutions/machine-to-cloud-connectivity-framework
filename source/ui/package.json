{
  "name": "machine-to-cloud-connectivity-framework-ui",
  "license": "Apache-2.0",
  "description": "Machine to Cloud Connectivity Framework UI",
  "author": {
    "name": "Amazon Web Services",
    "url": "https://aws.amazon.com/solutions"
  },
  "version": "4.2.0",
  "private": true,
  "devDependencies": {
    "@aws-amplify/api": "4.0.40",
    "@aws-amplify/auth": "4.5.4",
    "@aws-amplify/core": "4.5.4",
    "@aws-amplify/storage": "4.4.23",
    "@aws-amplify/ui-react": "2.17.1",
    "@testing-library/jest-dom": "^5.16.3",
    "@testing-library/react": "^12.1.4",
    "@testing-library/user-event": "^13.5.0",
    "@types/jest": "^27.4.1",
    "@types/react-dom": "^17.0.14",
    "@types/react-router-bootstrap": "^0.24.5",
    "@types/react-router-dom": "^5.3.3",
    "aws-amplify": "4.3.22",
    "bootstrap": "~4.6.0",
    "bootstrap-icons": "^1.8.1",
    "react": "^17.0.2",
    "react-bootstrap": "~1.6.4",
    "react-dom": "^17.0.2",
    "react-router-bootstrap": "^0.26.1",
    "react-router-dom": "^6.2.2",
    "react-scripts": "^5.0.0",
    "sass": "~1.49.9",
    "typescript": "~4.6.3"
  },
  "scripts": {
    "start": "react-scripts start",
    "clean": "rm -rf yarn.lock package-lock.json && rm -rf node_modules && rm -rf build",
    "build:react": "react-scripts build",
    "build": "yarn clean && yarn install && yarn build:react",
    "test": "react-scripts test --coverage --passWithNoTests --watchAll=false --silent",
    "eject": "react-scripts eject"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.{ts,tsx}",
      "!src/react-app-env.d.ts",
      "!src/index.tsx"
    ]
  }
}