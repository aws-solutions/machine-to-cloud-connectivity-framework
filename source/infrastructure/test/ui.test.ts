// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from 'aws-cdk-lib';
import { Bucket, CfnBucket } from 'aws-cdk-lib/aws-s3';
import { UiConstruct } from '../lib/frontend/ui';

test('M2C2 UI test', () => {
  const stack = new Stack();
  const s3LoggingBucket = new Bucket(stack, 'TestLoggingBucket');
  (<CfnBucket>s3LoggingBucket.node.defaultChild).overrideLogicalId('TestLoggingBucket');
  const resourceBucket = new Bucket(stack, 'TestGreengrassResourceBucket');
  (<CfnBucket>resourceBucket.node.defaultChild).overrideLogicalId('TestGreengrassResourceBucket');

  const ui = new UiConstruct(stack, 'TestUi', {
    apiId: 'mock-id',
    resourceBucket,
    userEmail: 'mockmail',
    cloudFrontDomainName: 'test-domain-name'
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  expect(ui.identityPoolId).toBeDefined();
  expect(ui.userPoolId).toBeDefined();
  expect(ui.webClientId).toBeDefined();
  expect(stack).toHaveResourceLike('AWS::S3::Bucket', {
    CorsConfiguration: {
      CorsRules: [
        {
          AllowedHeaders: ['*'],
          AllowedMethods: ['GET'],
          AllowedOrigins: ['https://test-domain-name'],
          ExposedHeaders: ['ETag']
        }
      ]
    }
  });
});
