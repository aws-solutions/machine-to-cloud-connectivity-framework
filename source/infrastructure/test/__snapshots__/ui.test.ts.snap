// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`M2C2 UI test 1`] = `
Object {
  "Resources": Object {
    "TestGreengrassResourceBucket": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "CorsConfiguration": Object {
          "CorsRules": Array [
            Object {
              "AllowedHeaders": Array [
                "*",
              ],
              "AllowedMethods": Array [
                "GET",
              ],
              "AllowedOrigins": Array [
                "https://test-domain-name",
              ],
              "ExposedHeaders": Array [
                "ETag",
              ],
            },
          ],
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestLoggingBucket": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "TestUiIdentityPoolAuthenticatedRoleE51963FB": Object {
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-IAM5",
              "reason": "It does not allow wildcard permissions.",
            },
          ],
        },
      },
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "TestUiIdentityPoolF602C72D",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              " Identity Pool authenticated role",
            ],
          ],
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "execute-api:Invoke",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":execute-api:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":mock-id/prod/*",
                      ],
                    ],
                  },
                },
                Object {
                  "Action": "s3:GetObject",
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "TestGreengrassResourceBucket",
                            "Arn",
                          ],
                        },
                        "/public/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "ExecuteApiPolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestUiIdentityPoolF602C72D": Object {
      "Properties": Object {
        "AllowUnauthenticatedIdentities": false,
        "CognitoIdentityProviders": Array [
          Object {
            "ClientId": Object {
              "Ref": "TestUiUserPoolClient21325B1C",
            },
            "ProviderName": Object {
              "Fn::GetAtt": Array [
                "TestUiUserPool04CDE8F0",
                "ProviderName",
              ],
            },
            "ServerSideTokenCheck": false,
          },
        ],
      },
      "Type": "AWS::Cognito::IdentityPool",
    },
    "TestUiIdentityPoolRoleAttachment72A6FD6E": Object {
      "Properties": Object {
        "IdentityPoolId": Object {
          "Ref": "TestUiIdentityPoolF602C72D",
        },
        "Roles": Object {
          "authenticated": Object {
            "Fn::GetAtt": Array [
              "TestUiIdentityPoolAuthenticatedRoleE51963FB",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "TestUiUser09D84075": Object {
      "Properties": Object {
        "DesiredDeliveryMediums": Array [
          "EMAIL",
        ],
        "ForceAliasCreation": true,
        "UserAttributes": Array [
          Object {
            "Name": "email",
            "Value": "mockmail",
          },
          Object {
            "Name": "email_verified",
            "Value": "true",
          },
        ],
        "UserPoolId": Object {
          "Ref": "TestUiUserPool04CDE8F0",
        },
        "Username": "mockmail",
      },
      "Type": "AWS::Cognito::UserPoolUser",
    },
    "TestUiUserPool04CDE8F0": Object {
      "DeletionPolicy": "Delete",
      "Metadata": Object {
        "cdk_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "AwsSolutions-COG2",
              "reason": "No need to enable MFA as that is option for users.",
            },
          ],
        },
      },
      "Properties": Object {
        "AccountRecoverySetting": Object {
          "RecoveryMechanisms": Array [
            Object {
              "Name": "verified_phone_number",
              "Priority": 1,
            },
            Object {
              "Name": "verified_email",
              "Priority": 2,
            },
          ],
        },
        "AdminCreateUserConfig": Object {
          "AllowAdminCreateUserOnly": true,
          "InviteMessageTemplate": Object {
            "EmailMessage": "
          <p>
            You are invited to join Machine to Cloud Connectivity Framework.<br />
            https://test-domain-name
          </p>
          <p>
            Please sign in to Machine to Cloud Connectivity Framework using the temporary credentials below:<br />
            Username: <strong>{username}</strong><br />Password: <strong>{####}</strong>
          </p>
        ",
            "EmailSubject": "[Machine to Cloud Connectivity Framework] Login information",
          },
        },
        "AutoVerifiedAttributes": Array [
          "email",
        ],
        "EmailVerificationMessage": "The verification code to your new account is {####}",
        "EmailVerificationSubject": "Verify your new account",
        "Policies": Object {
          "PasswordPolicy": Object {
            "MinimumLength": 12,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": true,
            "RequireUppercase": true,
          },
        },
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolAddOns": Object {
          "AdvancedSecurityMode": "ENFORCED",
        },
        "UserPoolName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "-user-pool",
            ],
          ],
        },
        "UsernameAttributes": Array [
          "email",
        ],
        "VerificationMessageTemplate": Object {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "The verification code to your new account is {####}",
          "EmailSubject": "Verify your new account",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "TestUiUserPoolClient21325B1C": Object {
      "Properties": Object {
        "AllowedOAuthFlows": Array [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": Array [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": Array [
          "https://example.com",
        ],
        "ClientName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "AWS::StackName",
              },
              "-userpool-client",
            ],
          ],
        },
        "GenerateSecret": false,
        "PreventUserExistenceErrors": "ENABLED",
        "RefreshTokenValidity": 1440,
        "SupportedIdentityProviders": Array [
          "COGNITO",
        ],
        "TokenValidityUnits": Object {
          "RefreshToken": "minutes",
        },
        "UserPoolId": Object {
          "Ref": "TestUiUserPool04CDE8F0",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
  },
}
`;
